service/lib/agama/registration.rb

```diff
@@ -0,0 +1,173 @@
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "fileutils"
+require "yast"
+require "ostruct"
+require "suse/connect"
+require "y2packager/new_repository_setup"
+
+Yast.import "Arch"
+
+module Agama
+  # Handles everything related to registration of system to SCC, RMT or similar
+  class Registration
+    attr_reader :reg_code
+
+    attr_reader :email
+
+    module Requirement
+      NOT_REQUIRED = :not_required
+      OPTIONAL = :optional
+      MANDATORY = :mandatory
+    end
+
+    # initializes registration with instance of software manager for query about products
+    def initialize(software_manager, logger)
+      @software = software_manager
+      @logger = logger
+    end
+
+    def register(code, email: "")
+      return unless product
+
+      connect_params = {
+        token: code,
+        email: email
+      }
+
+      login, password = SUSE::Connect::YaST.announce_system(connect_params, target_distro)
+      # write the global credentials
+      # TODO: check if we can do it in memory for libzypp
+      SUSE::Connect::YaST.create_credentials_file(login, password)
+
+      # TODO: fill it properly for scc
+      target_product = OpenStruct.new(
+        arch:       Yast::Arch.rpm_arch,
+        identifier: product.id,
+        version:    product.version
+      )
+      activate_params = {}
+      @service = SUSE::Connect::YaST.activate_product(target_product, activate_params, email)
+      Y2Packager::NewRepositorySetup.instance.add_service(@service.name)
+      add_service(@service)
+
+      @reg_code = code
+      @email = email
+      run_on_change_callbacks
+    end
+
+    def deregister
+      Y2Packager::NewRepositorySetup.instance.services.delete(@service.name)
+      remove_service(@service)
+
+      connect_params = {
+        token: reg_code,
+        email: email
+      }
+      SUSE::Connect::YaST.deactivate_system(connect_params)
+      FileUtils.rm(SUSE::Connect::YaST::GLOBAL_CREDENTIALS_FILE) # connect does not remove it itself
+
+      # reset varibles here
+      @reg_code = nil
+      @email = nil
+      run_on_change_callbacks
+    end
+
+    def requirement
+      return Requirement::NOT_REQUIRED unless product
+      return Requirement::MANDATORY if product.repositories.none?
+
+      Requirement::NOT_REQUIRED
+    end
+
+    # callback when state changed like when different product is selected
+    def on_change(&block)
+      @on_change_callbacks ||= []
+      @on_change_callbacks << block
+    end
+
+  private
+
+    attr_reader :software
+
+    def product
+      software.product
+    end
+
+    # E.g., "ALP-Dolomite-1-x86_64"
+    def target_distro
+      v = version.to_s.split(".").first || "1"
```

Comment:
here it crashes with unknown method version. I will try to fix it

Reply:
Thanks!